seps/1933-workload-identity-federation.md

```diff
@@ -0,0 +1,228 @@
+- The mapping between workload identity and MCP access token forms a critical trust boundary; misconfiguration can result in privilege escalation.
+- Administrators MUST configure issuer allowlists, claim-based authorization rules, and tenant isolation controls carefully.
+
+This SEP builds on established security models from OAuth 2.0, OIDC Discovery, and JWK-based key distribution. **Detailed and authoritative security guidance** is available in:
+
+- _RFC 7523_, Section 4 (Security Considerations for JWT Authorization Grants)
+- _RFC 7517_, Section 9 (Security Considerations for JWK)
+- _OpenID Connect Discovery 1.0_, Section 7 (Security Considerations)
+- The baseline MCP Authorization security requirements
+
+Implementers MUST review these documents when deploying Workload Identity Federation.
+
+The detailed proposal contains a complete list of security consideraitons (see https://github.com/modelcontextprotocol/ext-auth/blob/pieterkas-wif-extension/specification/draft/workload-identity-federation.mdx)
+
+## Reference Implementation
+
+A reference implementation is required before this SEP can achieve "Final" status. The reference implementation should demonstrate:
+
+1. **Authorization Server**: Token endpoint accepting JWT bearer grants with:
+   - OpenID Connect Discovery integration
+   - JWT validation using retrieved JWKs
+   - Configurable issuer and workload trust policies
+2. **MCP Client**: Workload client that:
+   - Obtains platform-issued JWT credentials
+   - Use this specification to present a JWT and receive an access token
+   - Uses access token for MCP server requests
+
+3. **Test Cases**: Including:
+   - Successful token issuence when a valid JWT is presented
+   - Rejection of untrusted issuers
+   - Rejection of unauthorized workloads
+   - JWT signature validation
+   - Replay attack prevention
+   - Multi-tenant isolation
+
+## Performance Implications
+
+The OpenID Connect Discovery process adds latency to the first token request for each issuer. However:
+
+- Issuer configurations and JWKs SHOULD be cached to minimize repeated discovery
+- Discovery overhead is incurred once per issuer, not per workload
+- The cost is comparable to other federated authentication mechanisms
+
+For high-throughput scenarios, authorization servers should implement:
+
+- Aggressive caching of issuer metadata and JWKs with appropriate TTLs
+- Connection pooling for issuer discovery requests
+- Pre-warming of cache for known trusted issuers
+
+## Open Questions
+
+1. **Token Refresh**: Should workload clients be allowed to obtain refresh tokens, or should they always present fresh workload JWTs for new access tokens?
+
+2. **Issuer Metadata Caching**: Should the specification provide guidance on cache TTLs for issuer configurations and JWKs?
+
+3. **Error Response Detail**: How much detail should authorization servers provide in error responses when JWT validation fails (for security vs. debuggability tradeoffs)?
```

Comment:
We're restricted somewhat by [RFC 7523](https://datatracker.ietf.org/doc/html/rfc7523#autoid-8). I might just call out that Authorization Servers SHOULD take that tradeoff into account.